docs/design/2021-03-30-small-table-caching.md

```diff
@@ -0,0 +1,174 @@
+
+
+A `CachedTable` struct will be introduced. It inherits and overloads the `Table` interface. `loadSchema` will periodically load the table information on schema change. For a cached table, the `CachedTable` will be reconstructed.
+
+`CachedTable` implements the `Table` interface, and overloads the `AddRecord` method. When `AddRecord` is called, the lock meta information needs to be checked first. This operation needs to acquire the "write" lock, that is to say, make the `lock_type` field of the meta table to 'WRITE'. After confirming that the lease time of the previous 'READ' lock has expired, the operation can be executed.
+
+The read operation flow diagram:
+
+![](./imgs/cached-table-read.png)
+
+
+The write operation flow diagram:
+
+![](./imgs/cached-table-write.png)
+
+
+### How to enable this feature
+
+The syntax of OceanBase's replicated table is to add the `DUPLICATE_SCOPE` option after the CREATE TABLE statement.
+
+I propose to use "ALTER TABLE t [CACHE|NOCACHE]" for the syntax. This is a switch that can be turned on or off. A user can decide to use it by themself. The write performance will become very bad after caching.
+
+"ALTER TABLE t CACHE" is a DDL operation. Assuming that all the TiDB instances know that a table is not cached, there is no correctness problem; Assuming that all tidb instances make a consensus that a table is cached, the correctness is guaranteed by the locking algorithm mentioned before. However, during the DDL operation, if some TiDB instances think that the cache is enabled, while some other instances think that the cache is not enabled, there would be a correctness issue. 
+
+To address that problem, an intermediate `Switching` state is introduced, the schema change process is similar to the LOCK TABLE statement's implementation:
+
+> Disabled => Switching => Enabled
+
+- In the Disabled state, all TiDB instances know that there is no cache for the table, it can be read and written normally.
+- In the Enabled state, all TiDB instances know that the table is cached, and the "lock" meta information should be checked before reading and writing.
+- In the switching state, all read operation goes to the original table; and write operation needs to check the "write lock" first.
+
+The cluster may have Disabled and Switching, or Switching and Enabled on different TiDB instances at the same time, but Disabled and Enabled will not coexist at the same time.
+
+### Compatibility
+
+The entry of the feature is the 'ALTER TABLE' statement. If the user does not explicitly call it, there will be no impact.
+
+Metadata management will introduce persistent data. The old version TiDB does not access or use the meta-data, so there should not be any problem when upgrading or downgrading.
+
+It is unsafe to mix the old and new versions of the binary, but rolling update is OK because there won’t be any cached tables in the old cluster. 
```

Comment:
The limitation is documented here @morgo

Reply:
I saw this, but it's not as clear to me as it should be. I don't think you will be able to solve it because it's blocked in Product Managment, but you can spell it out such as:

```suggestion
It is unsafe to mix the old and new versions of TiDB when using cached tables. Rolling updates are safe, but there is a risk that if an old version of a TiDB server re-joins after a table is in cached 'Enabled' state it could modify the contents in an unsafe manner. Currently _downgrade_ is not technically supported by TiDB, but there is no technical mechanism which prevents an older versioned binary joining a newer versioned cluster. We need to solve this issue to perform meta-data upgrades, but it is currently blocked in the requirements phase, because we have not decided from which versions upgrade/downgrade will be supported.
```